src/video/cocoa/cocoa_wnd.mm

```diff
@@ -343,156 +442,354 @@ - (void)appDidUnhide:(NSNotification*)note
+			HandleKeypress(pressed_key, unicode);
+		}
+		DEBUG(driver, 2, "cocoa_v: QZ_KeyEvent: %x (%x), down, mapping: %x", keycode, unicode, pressed_key);
+	} else {
+		DEBUG(driver, 2, "cocoa_v: QZ_KeyEvent: %x (%x), up", keycode, unicode);
+	}
+
+	return interpret_keys;
+}
+
+- (void)keyDown:(NSEvent *)event
+{
+	/* Quit, hide and minimize */
+	switch (event.keyCode) {
+		case QZ_q:
+		case QZ_h:
+		case QZ_m:
+			if (event.modifierFlags & NSCommandKeyMask) {
+				[ self interpretKeyEvents:[ NSArray arrayWithObject:event ] ];
+			}
+			break;
+	}
+
+	/* Convert UTF-16 characters to UCS-4 chars. */
+	std::vector<WChar> unicode_str = NSStringToUTF32([ event characters ]);
+	if (unicode_str.empty()) unicode_str.push_back(0);
+
+	if (EditBoxInGlobalFocus()) {
+		if ([ self internalHandleKeycode:event.keyCode unicode:unicode_str.front() pressed:YES modifiers:event.modifierFlags ]) {
+			[ self interpretKeyEvents:[ NSArray arrayWithObject:event ] ];
+		}
+	} else {
+		[ self internalHandleKeycode:event.keyCode unicode:unicode_str[0] pressed:YES modifiers:event.modifierFlags ];
+		for (size_t i = 1; i < unicode_str.size(); i++) {
+			[ self internalHandleKeycode:0 unicode:unicode_str[i] pressed:YES modifiers:event.modifierFlags ];
+		}
+	}
+}
+
+- (void)keyUp:(NSEvent *)event
+{
+	/* Quit, hide and minimize */
+	switch (event.keyCode) {
+		case QZ_q:
+		case QZ_h:
+		case QZ_m:
+			if (event.modifierFlags & NSCommandKeyMask) {
+				[ self interpretKeyEvents:[ NSArray arrayWithObject:event ] ];
+			}
+			break;
+	}
+
+	/* Convert UTF-16 characters to UCS-4 chars. */
+	std::vector<WChar> unicode_str = NSStringToUTF32([ event characters ]);
+	if (unicode_str.empty()) unicode_str.push_back(0);
+
+	[ self internalHandleKeycode:event.keyCode unicode:unicode_str[0] pressed:NO modifiers:event.modifierFlags ];
+}
+
+- (void)flagsChanged:(NSEvent *)event
+{
+	const int mapping[] = { QZ_CAPSLOCK, QZ_LSHIFT, QZ_LCTRL, QZ_LALT, QZ_LMETA };
+
+	NSUInteger newMods = event.modifierFlags;
+	if (self->_current_mods == newMods) return;
+
+	/* Iterate through the bits, testing each against the current modifiers */
+	for (unsigned int i = 0, bit = NSAlphaShiftKeyMask; bit <= NSCommandKeyMask; bit <<= 1, ++i) {
+		unsigned int currentMask, newMask;
+
+		currentMask = self->_current_mods & bit;
+		newMask     = newMods & bit;
+
+		if (currentMask && currentMask != newMask) { // modifier up event
+			/* If this was Caps Lock, we need some additional voodoo to make SDL happy (is this needed in ottd?) */
+			if (bit == NSAlphaShiftKeyMask) [ self internalHandleKeycode:mapping[i] unicode:0 pressed:YES modifiers:newMods ];
+			[ self internalHandleKeycode:mapping[i] unicode:0 pressed:NO modifiers:newMods ];
+		} else if (newMask && currentMask != newMask) { // modifier down event
+			[ self internalHandleKeycode:mapping[i] unicode:0 pressed:YES modifiers:newMods ];
+			/* If this was Caps Lock, we need some additional voodoo to make SDL happy (is this needed in ottd?) */
```

Comment:
What's SDL got to do with anything here?

Reply:
Most of the video driver was initially replicating SDL.

But, yes, that comment doesn't make much sense. Just like maybe the code line the comments refers to.